sycl/test-e2e/Matrix/joint_matrix_transposeC_impl.hpp

```diff
@@ -1,69 +1,129 @@
-           //  M x N => TM x N => TM x TN => TN x TM
-           //           m=sg_startx
-           //                      sg_starty/SG_SZ
-           // linear_index = M * (sg_starty/SG_SZ *TN) + TM *sg_startx
-           joint_matrix_load(sg, sub_c,
-                             pC + M * (sg_starty / SG_SZ * TN) + TM * sg_startx,
-                             M, layout::col_major);
-           joint_matrix_store(
-               sg, sub_c, pC + M * (sg_starty / SG_SZ * TN) + TM * sg_startx, M,
-               layout::col_major);
+           joint_matrix<sub_group, float, use::accumulator, TM, TN> sub_matrix;
+
+           auto row_major_offset =
+               (sg_startx * TM) * N + (sg_starty / SG_SZ * TN);
+           auto col_major_offset =
+               (sg_startx * TM) + (sg_starty / SG_SZ * TN) * M;
+
+           joint_matrix_load(sg, sub_matrix, p_input + col_major_offset, N,
+                             layout::col_major);
+
+           joint_matrix_store(sg, sub_matrix,
+                              p_out_col_major + row_major_offset, N,
+                              layout::row_major);
+
+           joint_matrix_store(sg, sub_matrix,
+                              p_out_row_major + col_major_offset, N,
+                              layout::col_major);
          }); // parallel for
    }).wait();
 }
 
-int main() {
+template <size_t TM> bool run_matrix_test() {
   static constexpr size_t MATRIX_M = 1024;
   static constexpr size_t MATRIX_N = 1024;
 
   queue q;
-  float *C = malloc_shared<float>(MATRIX_M * MATRIX_N, q);
-  float *D = malloc_shared<float>(MATRIX_M * MATRIX_N, q);
+  float *input = malloc_shared<float>(MATRIX_M * MATRIX_N, q);
+  float *out_col_major = malloc_shared<float>(MATRIX_M * MATRIX_N, q);
+  float *out_row_major = malloc_shared<float>(MATRIX_M * MATRIX_N, q);
+  float *ref_col_major = malloc_shared<float>(MATRIX_M * MATRIX_N, q);
 
-  matrix_rand(MATRIX_M, MATRIX_N, C, (float)5.0);
-  matrix_copy(MATRIX_M, MATRIX_N, C, D);
+  matrix_rand(MATRIX_M, MATRIX_N, input, (float)5.0);
+  matrix_fill(MATRIX_M, MATRIX_N, out_col_major, (float)0);
+  matrix_fill(MATRIX_M, MATRIX_N, out_row_major, (float)0);
+  matrix_transpose(MATRIX_M, MATRIX_N, ref_col_major, input);
 
-  matrix_load_store<float, MATRIX_M, MATRIX_N>(C, q);
+  matrix_load_and_store<TM, TN, float, MATRIX_M, MATRIX_N>(input, out_col_major,
+                                                           out_row_major, q);
 
-  bool res = matrix_compare(MATRIX_M, MATRIX_N, C, D);
+  // we use exact comparison as no low precision calculation is used in this
+  // test
+  bool res =
+      matrix_compare_exact(MATRIX_M, MATRIX_N, out_col_major, ref_col_major) &&
+      matrix_compare_exact(MATRIX_M, MATRIX_N, out_row_major, input);
+
+  free(input, q);
+  free(out_col_major, q);
+  free(out_row_major, q);
+  free(ref_col_major, q);
+  return res;
+}
+
+int main() {
+  bool res = true;
+  if (res)
+    res = run_matrix_test<8>();
+  if (res)
```

Comment:
Since we want to stop after the first incorrect result, I think in this case using assert will give shorter code.
`bool run_matrix_test` will become `void run_matrix_test` and instead of `return res` you can use `assert`
Then the code in main becomes just:
```
run_matrix_test<8>();
run_matrix_test<7>();
run_matrix_test<6>();
...
return 0;
```